### ISSUE TITLE
FatalExecutionEngineError in managed code

### ISSUE THREAD

**Issue by jptrxff** [dotnet/coreclr#20690]
Hi, 

I have experienced Access Violation Exceptions and Fatal Execution Engine Error (it seems to be the same error) in managed code in C#. I have narrowed it down to the following snippet. Am I missing something? I thought it should not be possible to these kind of exceptions in managed code. I have seen it both in .net 4.7 and 4.5

```c#
public static class FatalExecutionEngineBugExposer
    {

        public static void Main(string[] args)
        {

            for (int i = 0; i < 500000; i++)
            {
                try
                {
                    var testProblem = new TestProblem();
                    testProblem.AddTrianglesExperiment();
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.GetBaseException());
                }
            }
        }

        public class TestProblem
        {
            public struct Point
            {
                public double X, Y, Z;
            }

            public void AddTrianglesExperiment()
            {
                var points = new Point[28800];

                Parallel.ForEach(Partitioner.Create(0, points.Length),
                    range =>
                    {
                        // Create cache
                        var cache = new CubeRefCache();
                        cache.Entries = new CubeRefCacheEntry[20000];

                        // Add triangles
                        for (int i = range.Item1; i < range.Item2; i++)
                        {
                            ProcessTriangle(ref points[i].X, ref points[i].Y, ref points[i].Z, cache);  
                        }
                    });
            }
            
            void ProcessTriangle(ref double pt1, ref double pt2, ref double pt3, CubeRefCache cache)
            {
                cache.Add(0, 0);
            }

            public struct CubeRefCacheEntry
            {
                public int Index;
                public int Item;
            }
            
            class CubeRefCache
            {
                internal CubeRefCacheEntry[] Entries;

                internal void Add(int index, int item)
                {
                    Entries[0].Index = index;
                    Entries[0].Item = item;
                }
            }
        }
    }
```

**Comment by leculver** [comment]
We normally don't accept Desktop .Net issues through this repo.  I assume someone from the team who works here more often will come and either close this issue or correct me if I'm wrong.  =)

With that said, I agree we should never crash with an AV on something like what you've posted here, so I grabbed your repro and ran it in a hard loop all day (and will let it run overnight as well).  I've not seen any crash on CLR build 4.7.3190.0.

If you are running this code on a build before 4.7.3190.0, it's possible that there was a bug and it's fixed.  If your "4.7" run was on this version or later, I suspect you might be hitting a hardware issue.  Maybe bad ram or something related.

**Comment by jptrxff** [comment]
Thank you. If this is the wrong place, can you then direct me to where I should post this?

It seems that the issue is primarily seen in x64, and has been reproduced on at least 5 different pc's. I also posted this issue on stack overflow, where other users where able to reproduce it. 

https://stackoverflow.com/questions/53059965/fatal-execution-engine-error-in-managed-c-sharp-code

I have tried to run it on 4.7.2 as the latest, where it still fails.

**Comment by 0xd4d** [comment]
This repros in Debug + x64 on .NET Framework 4.7.2 but not on .NET Core 2.1 x64.

It looks like it's the GC that interrupts the thread after it has copied `cache` from `[rbp-XXX]` to `[rsp+20h]` in the delegate. `[rbp-XXX]` gets updated but not `[rsp+20h]` causing the called method to get a reference to the old `cache` on the GC heap.

.NET Core 2.1's jitter generates slightly different code and only copies it to `[rsp+20h]` just before calling the method. I don't know if it does that because it's a bug fix or if it's just lucky.

Workaround \#1:

Update app.config and disable BGC. I didn't get an exception, but I didn't run it for hours either.

```xml
<configuration>
	<startup>
		<supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.7.2" />
	</startup>
	<runtime>
		<gcConcurrent enabled="false"/>
	</runtime>
</configuration>
```

Workaround \#2:

Use Release builds + no debugger. Seems like it's less likely to happen.

Workaround \#3:

Add a dummy method that returns the input value. This could be difficult if there are a lot of calls to similar methods:

```C#
...
	ProcessTriangle(ref points[i].X, ref points[i].Y, ref points[i].Z, GetInputArg(cache));
...
	static T GetInputArg<T>(T value) => value;
...
```

**Comment by AndyAyersMS** [comment]
Seems likely this is #16892 / #19651.
cc @dotnet/jit-contrib 

@briansull did we bring the fix for this to 4.7 servicing?

**Comment by briansull** [comment]
Fixed in 4.8
